okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientStream.java

```diff
@@ -139,32 +140,33 @@ void onStreamSentBytes(int numBytes) {
     onSentBytes(numBytes);
   }
 
+  /**
+   * Must be called with holding the transport lock.
+   */
   public void transportHeadersReceived(List<Header> headers, boolean endOfStream) {
-    synchronized (lock) {
-      if (endOfStream) {
-        transportTrailersReceived(Utils.convertTrailers(headers));
-      } else {
-        transportHeadersReceived(Utils.convertHeaders(headers));
-      }
+    if (endOfStream) {
+      transportTrailersReceived(Utils.convertTrailers(headers));
+    } else {
+      transportHeadersReceived(Utils.convertHeaders(headers));
     }
   }
 
-  /**
-   * We synchronized on "lock" for delivering frames and updating window size, because
-   * the {@link #request(int)} call can be called in other thread for delivering frames.
-   */
   public void transportDataReceived(okio.Buffer frame, boolean endOfStream) {
+    boolean windowExceeded = false;
     synchronized (lock) {
       long length = frame.size();
       window -= length;
-      if (window < 0) {
-        frameWriter.rstStream(id(), ErrorCode.FLOW_CONTROL_ERROR);
-        Status status = Status.INTERNAL.withDescription(
-            "Received data size exceeded our receiving window size");
-        transport.finishStream(id(), status, null);
-        return;
+      windowExceeded = window < 0;
+      if (!windowExceeded) {
+        super.transportDataReceived(new OkHttpReadableBuffer(frame), endOfStream);
       }
-      super.transportDataReceived(new OkHttpReadableBuffer(frame), endOfStream);
+    }
+
+    if (windowExceeded) {
+      frameWriter.rstStream(id(), ErrorCode.FLOW_CONTROL_ERROR);
```

Comment:
This is an error case, it will be pretty fast, and we already handle the normal case with the lock held. Could we just handle this case with the lock held as well?

In fact, it seems we are going to need to hold the lock anyway within `finishStream()` and should likely be holding the lock between `rstStream()` and `finishStream()`.

Basically, I'm not convinced this _can_ be moved out of the synchronized block and I see many reasons why moving it out doesn't even help.